pinot/generative/torch_gvae/layers.py

```diff
@@ -0,0 +1,34 @@
+import torch
+import torch.nn.functional as F
+from torch.nn.modules.module import Module
+from torch.nn.parameter import Parameter
+
+
+class GraphConvolution(Module):
```

Comment:
I think this bit is a little repetitive. The graph -> latent code transform is taken care of by `pinot.representation` with API to dgl models and so on. Now it looks like this is not too much different than a vanilla GN but if you feel the need to implement new representation architecture maybe we should do it in `pinot.representation`

Reply:
I'd be hgappy to wait with fusing pinot representation with this for a while.
Ultimately yes, we want the same things.
For now, I just would like some empirical feedback on how well these different unsupervised choices work without being bogged down by pinot engineering.

Reply:
I thought that a lot of the code was somewhat repetitive too but it just happens to be this way because one half of the GVAE (encoder part) is essentially doing what `representation` does. Therefore, I tried to make the interface for GVAE to closely match that of `Net`. The reason is that in the future, if we want to pretrain the `representation` component before jointly training with `regression`, we can do:

```
representation = GVAE(params_here)
train_unsupervised(representation)

output_regression = initialize_some_output_regression()
Net(representation, output_regression)
```

Reply:
The `.forward` method you ported from the reference repo isn't what defines a VGAE, or even the encoder part for that matter. The three-layer simple graph conv is chosen simply as one of the simplest kind of implementations of graph conv, or in general, a graph embedding (**g -> h**). What distinguishes these model, from my understanding, is the transform from latent space to a graph (usually an adjacency matrix + some attribute). (let's call it **h -> g**)

So I would strong suggest we only implement **h -> g** part of these generative models as **g -> h** could be easily replaced by our previously defined models.

In this case it seems that the `.forward` function could be replaced by

```
pinot.representation.Sequential(
    layer=pinot.representation.dgl_legacy.gn('GraphConv'),
    config=[ # some dimensions and activations ])
```

Reply:
Right, the `.forward` function from the reference repo has been renamed to `encode_and_decode`. `.foward` function in **this** implementation is doing what you're suggesting, mapping from **g-> h**. I was aware of this difference so I modified the interface so it matches more closely with what we expect from the interface of `Net`